### PR TITLE
Match update and event flag fix

### PR DESCRIPTION
**Pull request by YannickSoehngen**
Fixed GoodEventFlag, increased granularity from counter to Get4 level, added pulser Flag (previous part of GoodEventFlag) and implementation of Single-Sided-Matches

### CODE REVIEW

File: StRoot/StEvent/StETofHeader.h [hunk 1]
```diff
@@ -109,6 +111,7 @@ class StETofHeader : public StObject {
 
     std::vector< Bool_t > mMissMatchFlagVec;
     std::vector< Bool_t > mGoodEventFlagVec; 
+    std::vector< Bool_t > mHasPulsersVec; 
 
     ClassDef( StETofHeader, 3 )
```
Comment (klendathu2k) [inline StRoot/StEvent/StETofHeader.h]: Addition of mHasPulserVec on line 114 requires incrementing the argument to ClassDef.

File: StRoot/StMuDSTMaker/COMMON/StMuETofHeader.h [hunk 2]
```diff
@@ -115,7 +117,8 @@ class StMuETofHeader : public TObject {
     ULong64_t   mEventStatusFlag;
 
 	 std::vector< Bool_t > mMissMatchFlagVec;
-	 std::vector< Bool_t > mGoodEventFlagVec;  
+	 std::vector< Bool_t > mGoodEventFlagVec; 
+	 std::vector< Bool_t > mHasPulsersVec;
 
     ClassDef( StMuETofHeader, 3 )
```
Comment (klendathu2k) [inline StRoot/StMuDSTMaker/COMMON/StMuETofHeader.h]: Additional data field (line 121) requires incrementing the ClassDef.

File: StRoot/StPicoEvent/StPicoEvent.h [hunk 3]
```diff
@@ -609,8 +615,9 @@ class StPicoEvent : public TObject {
   UShort_t mETofHitMultiplicity ;
   /// Total digi multiplicity in ETOF modules
   UShort_t mETofDigiMultiplicity ;
-  /// Flag to mark if the event is good for physics analysis for each counter. A counter is considered good in each event when there are zero missmatch flags set and pulser digis on both sides are found. In this case, the counter should perform at its best. Counter efficiency should be constant between good events. Here: CounterNr = 9*sector + 3*module + counter.
-  bool mETofGoodEventFlag[108];
+  /// Flag to mark if the event is good for physics analysis for each Get4. A Get4 is considered good in each event when there are zero missmatch flags set.  Get4 efficiency should be constant between good events. As additional sanity check one can request that both pulsers were present for any given event and counter. Best performance to be expected with zero status bits and both pulsers. Here: CounterNr = 9*(sector-13) + 3*(module-1) + counter. Get4Nr = 3*3*16*(sector-13)+ 3*16*(module -1) + 16*(counter -1) + Get4.
+  bool mETofGoodEventFlag[1728];
+  bool mETofHasPulsersFlag[108];
```
Comment (klendathu2k) [inline StRoot/StPicoEvent/StPicoEvent.h:620]: ClassDef needs to be incremented.

**Comment by plexoos** [comment]
There was a small conflict with the main branch. I resolved it

File: StRoot/StETofCalibMaker/StETofCalibMaker.cxx [hunk 4]
```diff
@@ -1381,55 +1383,61 @@ StETofCalibMaker::processMuDst()
...
+		//drag along pulser information
+		for( int iCounter = 0; iCounter < 108; iCounter++){
+		  if ( !(mNPulsersCounter.count(iCounter) ) ){
+		    hasPulsersVec.push_back(false);
+		  }else{
+		    if(mNPulsersCounter[iCounter] == 2){
+		      hasPulsersVec.push_back(true);
+		    }else{
+		      hasPulsersVec.push_back(false);
+		    }
```
Comment (plexoos) [inline StRoot/StETofCalibMaker/StETofCalibMaker.cxx]:
```suggestion
		      hasPulsersVec.push_back(mNPulsersCounter[iCounter] == 2);    
```

File: StRoot/StETofCalibMaker/StETofCalibMaker.cxx [hunk 5]
```diff
@@ -1381,55 +1383,61 @@ StETofCalibMaker::processMuDst()
...
+		  etofHeader->setHasPulsersVec(hasPulsersVec);
+	  	} 
+
+		//fill good event flag into header
+		for( int iGet4 = 0; iGet4 < 1728; iGet4++){	  	
+		  if ( etofHeader->missMatchFlagVec().at(iGet4)){
+		    goodEventFlagVec.push_back(false); 
+		  }else{
+		    goodEventFlagVec.push_back(true);
+		  }
```
Comment (plexoos) [inline StRoot/StETofCalibMaker/StETofCalibMaker.cxx]:
```suggestion
		    goodEventFlagVec.push_back(!etofHeader->missMatchFlagVec().at(iGet4)); 
```

File: StRoot/StETofCalibMaker/StETofCalibMaker.cxx [hunk 6]
```diff
@@ -1233,34 +1233,35 @@ StETofCalibMaker::processStEvent()
...
+		  //etofHeader->setHasPulsersVec(hasPulsersVec);  // not working but not of relevance at the moment 
+	  	} 
+
+		//fill good event flag into header
+		for( int iGet4 = 0; iGet4 < 1728; iGet4++){	  	
+		  if ( etofHeader->missMatchFlagVec().at(iGet4)){
+		    goodEventFlagVec.push_back(false); 
+		  }else{
+		    goodEventFlagVec.push_back(true);
+		  }
```
Comment (plexoos) [inline StRoot/StETofCalibMaker/StETofCalibMaker.cxx]:
```suggestion
		    goodEventFlagVec.push_back( !etofHeader->missMatchFlagVec().at(iGet4) );
```

File: StRoot/StETofCalibMaker/StETofCalibMaker.cxx [hunk 7]
```diff
@@ -1233,34 +1233,35 @@ StETofCalibMaker::processStEvent()
...
+		//drag along pulser information
+		for( int iCounter = 0; iCounter < 108; iCounter++){
+		  if ( !(mNPulsersCounter.count(iCounter) ) ){
+		    hasPulsersVec.push_back(false);
+		  }else{
+		    if(mNPulsersCounter[iCounter] == 2){
+		      hasPulsersVec.push_back(true);
+		    }else{
+		      hasPulsersVec.push_back(false);
+		    }
```
Comment (plexoos) [inline StRoot/StETofCalibMaker/StETofCalibMaker.cxx]:
```suggestion
		      hasPulsersVec.push_back(mNPulsersCounter[iCounter] == 2);   
```

**Review by plexoos — COMMENTED** [review]
Please consider reducing the number of if-statements

**Comment by plexoos** [comment]
There are also errors from the compiler. Please address them

File: StRoot/StETofCalibMaker/StETofCalibMaker.cxx (hunk 6 above)
Comment (plexoos) [inline StRoot/StETofCalibMaker/StETofCalibMaker.cxx]: This should be re-written in modern C++ without using an explicit loop. The both source and destination vectors appear to contain boolean values, so using std::copy with back_inserter should work.

**Review by klendathu2k — APPROVED** [review]
StEvent etc... looks good now.

**Review by plexoos — APPROVED** [review]

**Comment by genevb** [comment]
Are we waiting on anything here? A review from @PhilippWeidenkaff ? Or is this ready to merge?

**Review by fgeurts — APPROVED** [review]

**Comment by genevb** [comment]
I understand that this is the only PR on which we are critically waiting before tagging SL24a. I'll ask again: are we waiting for anything else on this PR before merging?

**Comment by genevb** [comment]
Nightly tests look like they went OK. @plexoos , would you please tag SL24a here and `star-mcgen` when you get the chance? Thanks in advance.

**Comment by plexoos** [comment]
Done. Branch SL24a and tag SL24a_0 should be available in the both repositories now

**Pull request merged**